tracer/src/Datadog.Trace/Ci/CIVisibility.cs

```diff
@@ -0,0 +1,217 @@
+
+        public static void Initialize()
+        {
+            if (Interlocked.Exchange(ref _firstInitialization, 0) != 1)
+            {
+                // Initialize() was already called before
+                return;
+            }
+
+            Log.Information("Initializing CI Visibility");
+
+            // Set exit handlers
+            AppDomain.CurrentDomain.ProcessExit += CurrentDomain_ProcessExit;
+            AppDomain.CurrentDomain.DomainUnload += CurrentDomain_DomainUnload;
+            try
+            {
+                // Registering for the AppDomain.UnhandledException event cannot be called by a security transparent method
+                // This will only happen if the Tracer is not run full-trust
+                AppDomain.CurrentDomain.UnhandledException += CurrentDomain_UnhandledException;
+            }
+            catch (Exception ex)
+            {
+                Log.Warning(ex, "Unable to register a callback to the AppDomain.UnhandledException event.");
+            }
+
+            try
+            {
+                // Registering for the cancel key press event requires the System.Security.Permissions.UIPermission
+                Console.CancelKeyPress += Console_CancelKeyPress;
+            }
+            catch (Exception ex)
+            {
+                Log.Warning(ex, "Unable to register a callback to the Console.CancelKeyPress event.");
+            }
+
+            TracerSettings tracerSettings = _settings.TracerSettings ?? TracerSettings.FromDefaultSources();
+
+            // Set the tracer buffer size to the max
+            tracerSettings.TraceBufferSize = 1024 * 1024 * 45; // slightly lower than the 50mb payload agent limit.
+
+            // Set the service name if empty
+            Log.Information("Setting up the service name");
+            if (string.IsNullOrEmpty(tracerSettings.ServiceName))
+            {
+                // Extract repository name from the git url and use it as a default service name.
+                string repository = CIEnvironmentValues.Repository;
+                if (!string.IsNullOrEmpty(repository))
+                {
+                    Regex regex = new Regex(@"/([a-zA-Z0-9\\\-_.]*)$");
+                    Match match = regex.Match(repository);
+                    if (match.Success && match.Groups.Count > 1)
+                    {
+                        const string gitSuffix = ".git";
+                        string repoName = match.Groups[1].Value;
+                        if (repoName.EndsWith(gitSuffix))
+                        {
+                            tracerSettings.ServiceName = repoName.Substring(0, repoName.Length - gitSuffix.Length);
+                        }
+                        else
+                        {
+                            tracerSettings.ServiceName = repoName;
+                        }
+                    }
+                }
+            }
```

Comment:
Feels like you might want to extract that and wrap a couple of unit tests around it, make sure we're handling the edge cases etc?

Reply:
Yeah, that makes sense, will do.

Reply:
done in d854823